"invalid memory address or nil..." in the scheduler on flynn scale

Today I did a `flynn scale web=10` immediately followed by a `flynn scale web=0`, and it caused scheduler to go down.  Unsure of an exact way to reproduce this.  It was with master of the project built from source, and it was on the very first run of the cluster.  Here is the output from the scheduler right before it went down:

```
2014/06/23 23:31:28 rpc: client protocol error: unexpected EOF
2014/06/23 23:31:28 rpc: client protocol error: unexpected EOF
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x40 pc=0x4070a1]

goroutine 156 [running]:
runtime.panic(0x6c2440, 0xa63b28)
    /usr/local/go/src/pkg/runtime/panic.c:266 +0xb6
main.(*Formation).remove(0xc2100eb300, 0xa, 0xc2100f3c38, 0x3)
    /vagrant/src/flynn-controller/scheduler/main.go:619 +0x3e1
main.(*Formation).rectify(0xc2100eb300)
    /vagrant/src/flynn-controller/scheduler/main.go:521 +0x72b
main.(*Formation).Rectify(0xc2100eb300)
    /vagrant/src/flynn-controller/scheduler/main.go:463 +0x5f
main.func·001()
    /vagrant/src/flynn-controller/scheduler/main.go:190 +0x37
created by main.(*context).watchFormations
    /vagrant/src/flynn-controller/scheduler/main.go:194 +0x43d

goroutine 1 [chan receive]:
main.(*context).watchFormations(0xc210022e10, 0x0)
    /vagrant/src/flynn-controller/scheduler/main.go:175 +0x1a0
main.main()
    /vagrant/src/flynn-controller/scheduler/main.go:47 +0x5d9

goroutine 4 [IO wait]:
net.runtime_pollWait(0x7fab2af4d5e8, 0x72, 0x0)
    /usr/local/go/src/pkg/runtime/netpoll.goc:116 +0x6a
net.(*pollDesc).Wait(0xc210055ca0, 0x72, 0x7fab2af4c170, 0xb)
    /usr/local/go/src/pkg/net/fd_poll_runtime.go:81 +0x34
net.(*pollDesc).WaitRead(0xc210055ca0, 0xb, 0x7fab2af4c170)
    /usr/local/go/src/pkg/net/fd_poll_runtime.go:86 +0x30
net.(*netFD).Read(0xc210055c40, 0xc210074000, 0x1000, 0x1000, 0x0, ...)
    /usr/local/go/src/pkg/net/fd_unix.go:204 +0x2a0
net.(*conn).Read(0xc210000370, 0xc210074000, 0x1000, 0x1000, 0x7fab2af43000, ...)
    /usr/local/go/src/pkg/net/net.go:122 +0xc5
bufio.(*Reader).fill(0xc210041480)
    /usr/local/go/src/pkg/bufio/bufio.go:91 +0x110
bufio.(*Reader).Read(0xc210041480, 0xc21000acb0, 0x1, 0x9, 0x1, ...)
    /usr/local/go/src/pkg/bufio/bufio.go:159 +0x1a4
io.ReadAtLeast(0x7fab2af4d7c8, 0xc210041480, 0xc21000acb0, 0x1, 0x9, ...)
    /usr/local/go/src/pkg/io/io.go:288 +0xf6
io.ReadFull(0x7fab2af4d7c8, 0xc210041480, 0xc21000acb0, 0x1, 0x9, ...)
    /usr/local/go/src/pkg/io/io.go:306 +0x71
encoding/gob.decodeUintReader(0x7fab2af4d7c8, 0xc210041480, 0xc21000acb0, 0x9, 0x9, ...)
    /usr/local/go/src/pkg/encoding/gob/decode.go:66 +0x98
encoding/gob.(*Decoder).recvMessage(0xc21000b0f0, 0xc21003c270)
    /usr/local/go/src/pkg/encoding/gob/decoder.go:73 +0x57
encoding/gob.(*Decoder).decodeTypeSequence(0xc21000b0f0, 0xc21006df00, 0x160)
    /usr/local/go/src/pkg/encoding/gob/decoder.go:159 +0x49
encoding/gob.(*Decoder).DecodeValue(0xc21000b0f0, 0x6238a0, 0xc21006df90, 0x160, 0x0, ...)
    /usr/local/go/src/pkg/encoding/gob/decoder.go:223 +0x12e
encoding/gob.(*Decoder).Decode(0xc21000b0f0, 0x6238a0, 0xc21006df90, 0xc210057218, 0xc2100d42a0)
    /usr/local/go/src/pkg/encoding/gob/decoder.go:202 +0x1c5
github.com/flynn/rpcplus.(*gobClientCodec).ReadResponseHeader(0xc21006ddb0, 0xc21006df90, 0xc2100782d0, 0x0)
    /tmp/godep/rev/05/df3d30b6e55f5b440c1a78e2f83cb822bc0d76/src/github.com/flynn/rpcplus/client.go:278 +0x3f
github.com/flynn/rpcplus.(*Client).input(0xc2100414e0)
    /tmp/godep/rev/05/df3d30b6e55f5b440c1a78e2f83cb822bc0d76/src/github.com/flynn/rpcplus/client.go:138 +0xb8
created by github.com/flynn/rpcplus.NewClientWithCodec
    /tmp/godep/rev/05/df3d30b6e55f5b440c1a78e2f83cb822bc0d76/src/github.com/flynn/rpcplus/client.go:256 +0x91

goroutine 5 [chan receive]:
github.com/flynn/go-discoverd.func·001()
    /tmp/godep/rev/ef/8e9c70ed6335e8db5a66addfa96226f35e2bd3/src/github.com/flynn/go-discoverd/client.go:123 +0x5b
created by github.com/flynn/go-discoverd.(*serviceSet).bind
    /tmp/godep/rev/ef/8e9c70ed6335e8db5a66addfa96226f35e2bd3/src/github.com/flynn/go-discoverd/client.go:165 +0xcf

goroutine 6 [chan receive]:
github.com/flynn/go-flynn/cluster.(*Client).followLeader(0xc2100573f0, 0x0)
    /tmp/godep/rev/77/5d010f4f7ff91b144418ad74c6f6e285d50f49/src/github.com/flynn/go-flynn/cluster/client.go:90 +0x50
created by github.com/flynn/go-flynn/cluster.(*Client).start
    /tmp/godep/rev/77/5d010f4f7ff91b144418ad74c6f6e285d50f49/src/github.com/flynn/go-flynn/cluster/client.go:85 +0x53

goroutine 7 [chan receive]:
github.com/flynn/go-discoverd.func·002()
    /tmp/godep/rev/ef/8e9c70ed6335e8db5a66addfa96226f35e2bd3/src/github.com/flynn/go-discoverd/client.go:224 +0x8d
created by github.com/flynn/go-discoverd.(*serviceSet).Leaders
    /tmp/godep/rev/ef/8e9c70ed6335e8db5a66addfa96226f35e2bd3/src/github.com/flynn/go-discoverd/client.go:230 +0xea

goroutine 9 [select]:
github.com/flynn/go-discoverd.func·005()
    /tmp/godep/rev/ef/8e9c70ed6335e8db5a66addfa96226f35e2bd3/src/github.com/flynn/go-discoverd/client.go:491 +0x2e8
created by github.com/flynn/go-discoverd.(*Client).RegisterWithAttributes
    /tmp/godep/rev/ef/8e9c70ed6335e8db5a66addfa96226f35e2bd3/src/github.com/flynn/go-discoverd/client.go:502 +0x454

goroutine 10 [chan receive]:
github.com/flynn/go-discoverd.func·001()
    /tmp/godep/rev/ef/8e9c70ed6335e8db5a66addfa96226f35e2bd3/src/github.com/flynn/go-discoverd/client.go:123 +0x5b
created by github.com/flynn/go-discoverd.(*serviceSet).bind
    /tmp/godep/rev/ef/8e9c70ed6335e8db5a66addfa96226f35e2bd3/src/github.com/flynn/go-discoverd/client.go:165 +0xcf

goroutine 14 [chan receive]:
github.com/flynn/go-discoverd.func·002()
    /tmp/godep/rev/ef/8e9c70ed6335e8db5a66addfa96226f35e2bd3/src/github.com/flynn/go-discoverd/client.go:224 +0x8d
created by github.com/flynn/go-discoverd.(*serviceSet).Leaders
    /tmp/godep/rev/ef/8e9c70ed6335e8db5a66addfa96226f35e2bd3/src/github.com/flynn/go-discoverd/client.go:230 +0xea

goroutine 15 [chan receive]:
github.com/flynn/go-discoverd.func·001()
    /tmp/godep/rev/ef/8e9c70ed6335e8db5a66addfa96226f35e2bd3/src/github.com/flynn/go-discoverd/client.go:123 +0x5b
created by github.com/flynn/go-discoverd.(*serviceSet).bind
    /tmp/godep/rev/ef/8e9c70ed6335e8db5a66addfa96226f35e2bd3/src/github.com/flynn/go-discoverd/client.go:165 +0xcf

goroutine 16 [IO wait]:
net.runtime_pollWait(0x7fab2af4d498, 0x72, 0x0)
    /usr/local/go/src/pkg/runtime/netpoll.goc:116 +0x6a
net.(*pollDesc).Wait(0xc2100558b0, 0x72, 0x7fab2af4c170, 0xb)
    /usr/local/go/src/pkg/net/fd_poll_runtime.go:81 +0x34
net.(*pollDesc).WaitRead(0xc2100558b0, 0xb, 0x7fab2af4c170)
    /usr/local/go/src/pkg/net/fd_poll_runtime.go:86 +0x30
net.(*netFD).Read(0xc210055850, 0xc210087000, 0x1000, 0x1000, 0x0, ...)
    /usr/local/go/src/pkg/net/fd_unix.go:204 +0x2a0
net.(*conn).Read(0xc210000c90, 0xc210087000, 0x1000, 0x1000, 0x0, ...)
    /usr/local/go/src/pkg/net/net.go:122 +0xc5
bufio.(*Reader).fill(0xc2100812a0)
    /usr/local/go/src/pkg/bufio/bufio.go:91 +0x110
bufio.(*Reader).Read(0xc2100812a0, 0xc2100784d0, 0x1, 0x9, 0x1, ...)
    /usr/local/go/src/pkg/bufio/bufio.go:159 +0x1a4
io.ReadAtLeast(0x7fab2af4d7c8, 0xc2100812a0, 0xc2100784d0, 0x1, 0x9, ...)
    /usr/local/go/src/pkg/io/io.go:288 +0xf6
io.ReadFull(0x7fab2af4d7c8, 0xc2100812a0, 0xc2100784d0, 0x1, 0x9, ...)
    /usr/local/go/src/pkg/io/io.go:306 +0x71
encoding/gob.decodeUintReader(0x7fab2af4d7c8, 0xc2100812a0, 0xc2100784d0, 0x9, 0x9, ...)
    /usr/local/go/src/pkg/encoding/gob/decode.go:66 +0x98
encoding/gob.(*Decoder).recvMessage(0xc21000b3c0, 0xc21011f000)
    /usr/local/go/src/pkg/encoding/gob/decoder.go:73 +0x57
encoding/gob.(*Decoder).decodeTypeSequence(0xc21000b3c0, 0xc21007dc00, 0x160)
    /usr/local/go/src/pkg/encoding/gob/decoder.go:159 +0x49
encoding/gob.(*Decoder).DecodeValue(0xc21000b3c0, 0x6238a0, 0xc21007dc60, 0x160, 0x0, ...)
    /usr/local/go/src/pkg/encoding/gob/decoder.go:223 +0x12e
encoding/gob.(*Decoder).Decode(0xc21000b3c0, 0x6238a0, 0xc21007dc60, 0x0, 0x1)
    /usr/local/go/src/pkg/encoding/gob/decoder.go:202 +0x1c5
github.com/flynn/rpcplus.(*gobClientCodec).ReadResponseHeader(0xc21007dab0, 0xc21007dc60, 0x120, 0x622e60)
    /tmp/godep/rev/05/df3d30b6e55f5b440c1a78e2f83cb822bc0d76/src/github.com/flynn/rpcplus/client.go:278 +0x3f
github.com/flynn/rpcplus.(*Client).input(0xc210081300)
    /tmp/godep/rev/05/df3d30b6e55f5b440c1a78e2f83cb822bc0d76/src/github.com/flynn/rpcplus/client.go:138 +0xb8
created by github.com/flynn/rpcplus.NewClientWithCodec
    /tmp/godep/rev/05/df3d30b6e55f5b440c1a78e2f83cb822bc0d76/src/github.com/flynn/rpcplus/client.go:256 +0x91
```

So I looked into this, and it's occurring b/c if we pass in a non-existent job.HostId [here](https://github.com/flynn/flynn-controller/blob/master/scheduler/main.go#L619), it will return a nil value [here](https://github.com/flynn/flynn-controller/blob/master/scheduler/main.go#L313).  So the question becomes, do we simply want to check and see what Get() returns before calling stop on it?  If so, what do we do if it returns nil?  Just continue on?  Or is there some kind of failure case, b/c we don't want to leave the container up and running.

The problem may even stem further up the chain now that I think about it.  Check out the value of `f.c.hosts` as the error happens:

```
2014/06/24 03:59:46 f.c.hosts &{map[] {{0 0} 0 0 0 0}}
2014/06/24 03:59:46 f.c.hosts.Get(ID) <nil>
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x40 pc=0x407415]
```

--- a/lorne.go
+++ b/lorne.go
@@ -64,7 +64,7 @@ func main() {
 		log.Fatal(err)
 	}
 
-	go serveHTTP(&Host{state: state}, &attachHandler{state: state, backend: backend})
+	go serveHTTP(&Host{state: state, backend: backend}, &attachHandler{state: state, backend: backend})
 
 	if *force {
 		if err := backend.Cleanup(); err != nil {

